src/Components/ResourcePicker/ResourcePicker.tsx

```diff
@@ -0,0 +1,649 @@
+                return t('resourcesPicker.loadingContainers');
+            case AzureResourceTypes.StorageAccount:
+                return t('resourcesPicker.loadingStorageAccounts');
+            default:
+                return t('resourcesPicker.loadingResources');
+        }
+    }, [resourceType, t]);
+
+    const getDisplayFieldValue = useCallback(
+        (resource: IAzureResource) => {
+            if (displayField === AzureResourceDisplayFields.url) {
+                switch (resource.type) {
+                    case AzureResourceTypes.DigitalTwinInstance:
+                        return resource.properties.hostName
+                            ? 'https://' + resource.properties.hostName
+                            : null;
+                    case AzureResourceTypes.StorageAccount:
+                        return resource.properties.primaryEndpoints?.blob;
+                    default:
+                        return resource[
+                            AzureResourceDisplayFields[displayField] // storage blob containers does not have url information in its resource properties, it is constructed based on the storage account blob url
+                        ];
+                }
+            } else {
+                return resource[AzureResourceDisplayFields[displayField]];
+            }
+        },
+        [displayField]
+    );
+
+    // after fetching the resources merge them with existing options in the dropdown
+    const mergeOptionsWithProps = useCallback(
+        (options, additionalOptions, selectedOption) => {
+            if (additionalOptions) {
+                // do the merging with existing options: add the additional options manually entered by user to the end if it is not in the fetched data
+                const existingOptionTexts =
+                    resourcesRef.current?.map((resource) =>
+                        getDisplayFieldValue(resource)
+                    ) || [];
+
+                const optionsToAdd = additionalOptions.filter(
+                    (additionalOption) =>
+                        existingOptionTexts.findIndex((existingOptionText) =>
+                            isValuesEqual(
+                                existingOptionText,
+                                additionalOption.text
+                            )
+                        ) === -1
+                );
+
+                if (optionsToAdd.length) {
+                    if (resourcesRef.current?.length > 0) {
+                        // add the freeform options header if the resources are fetched and exists
+                        options.push(freeformOptionsHeader);
+                    }
+                    options = options.concat(optionsToAdd);
+                }
+            }
+            if (selectedOption) {
+                const selectedOptionInNewOptions = options.find((option) =>
+                    isValuesEqual(option.text, selectedOption.text)
```

Comment:
why do we find by matching the text instead of using a key/value property? Couldn't that lead to issues where we have manipulated a display value? Ex: adding https: or something to the start.

Reply:
the key value of the freeform options in the dropdown is the text itself (because there is no fetching resources for them). That is why I am comparing based on the text values we see in the dropdown to check if the options exists or not.

I dont understand what you mean by saying "a manipulated display value". In the dropdown there are options fetched from server, plus the manually added strings using freeform in the combobox. the manually added ones' key values are the text itself, and after adding an option there is no way to manipulate the value of it (you can only remove that option and add a new one)